server/src/player/digging.rs

```diff
@@ -236,6 +249,112 @@ fn handle_drop_item_stack(
     }
 }
 
+/// Handles food consumption and shooting arrows.
+fn handle_consume_item(
+    packet: &PlayerDigging,
+    player: &PlayerComponent,
+    entity: Entity,
+    inventory: &mut InventoryComponent,
+    inventory_updates: &mut EventChannel<InventoryUpdateEvent>,
+    position: Position,
+    shoot_arrow_events: &mut EventChannel<ShootArrowEvent>,
+) {
+    assert_eq!(packet.status, PlayerDiggingStatus::ConsumeItem);
+
+    // TODO: Fallback to off-hand if main-hand is not a consumable
+    let used_item = inventory.item_in_main_hand();
+
+    if let Some(item) = used_item {
+        if item.ty == Item::Bow {
+            handle_shoot_bow(
+                player,
+                entity,
+                inventory,
+                inventory_updates,
+                position,
+                shoot_arrow_events,
+            );
+        }
+        // TODO: Food, potions
+    }
+}
+
+fn handle_shoot_bow(
+    player: &PlayerComponent,
+    entity: Entity,
+    inventory: &mut InventoryComponent,
+    inventory_updates: &mut EventChannel<InventoryUpdateEvent>,
+    position: Position,
+    shoot_arrow_events: &mut EventChannel<ShootArrowEvent>,
+) {
+    let arrow_to_consume: Option<(SlotIndex, ItemStack)> = find_arrow(inventory.clone());
+    if player.gamemode == Gamemode::Survival || player.gamemode == Gamemode::Adventure {
+        // If no arrow was found, don't shoot
+        let arrow_to_consume = arrow_to_consume.clone();
+        if arrow_to_consume.is_none() {
+            debug!("Tried to shoot bow with no arrows.");
+            return;
+        }
+
+        // Consume arrow
+        let (arrow_slot, arrow_stack) = arrow_to_consume.unwrap();
+        let mut arrow_stack: ItemStack = arrow_stack.clone();
+        arrow_stack.amount -= 1;
+
+        inventory.set_item_at(arrow_slot, arrow_stack);
+        inventory_updates.single_write(InventoryUpdateEvent {
+            slots: smallvec![arrow_slot],
+            player: entity,
+        });
+    }
+
+    let arrow_type: Item = match arrow_to_consume {
+        None => Item::Arrow, // Default to generic arrow in creative mode with none in inventory
+        Some((_, arrow_stack)) => arrow_stack.ty,
+    };
+
+    shoot_arrow_events.single_write(ShootArrowEvent {
+        shooter: Some(entity),
+        position,
+        arrow_type,
+        critical: false, // TODO: Determine critical based on how long bow was pulled back
+    });
+}
+
+fn find_arrow(inventory: InventoryComponent) -> Option<(SlotIndex, ItemStack)> {
```

Comment:
Passing a reference to `InventoryComponent` should suffice here. No need to clone the entire inventory, just clone the item stacks you retrieve.